packages/php-wasm/compile/build-assets/php_wasm.c

```diff
@@ -111,7 +111,15 @@ PHP_FUNCTION(post_message_to_js)
 		return;
 	}
 
-	js_module_onMessage(data);
+	char *result = js_module_onMessage(data);
+
+	#if PHP_MAJOR_VERSION >= 8
```

Comment:
Ran into problem trying to compile this for other PHP versions, I need to check for which other ones this fails. I guess we could just not supported returning data for older versions.

Reply:
what was the error it gave you?

Reply:
I think the main error was:

```
0.308 /root/php_wasm.c:117:3: error: use of undeclared identifier 'RETURN_STRING'
0.308   117 |                 RETURN_STRING(result);
```

Here's a larger piece of the log:

```
#90 0.165 Setting up EMSDK environment (suppress these messages with EMSDK_QUIET=1)
#90 0.165 Adding directories to PATH:
#90 0.165 PATH += /root/emsdk
#90 0.165 PATH += /root/emsdk/upstream/emscripten
#90 0.165 PATH += /root/emsdk/node/16.20.0_64bit/bin
#90 0.165 
#90 0.165 Setting environment variables:
#90 0.165 PATH = /root/emsdk:/root/emsdk/upstream/emscripten:/root/emsdk/node/16.20.0_64bit/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin
#90 0.165 EMSDK = /root/emsdk
#90 0.165 EMSDK_NODE = /root/emsdk/node/16.20.0_64bit/bin/node
#90 0.300 In file included from /root/php_wasm.c:13:
#90 0.300 In file included from ./ext/standard/php_standard.h:43:
#90 0.300 In file included from ./ext/standard/fsock.h:30:
#90 0.300 In file included from main/php_network.h:112:
#90 0.300 /root/emsdk/upstream/emscripten/cache/sysroot/include/sys/poll.h:1:2: warning: redirecting incorrect #include <sys/poll.h> to <poll.h> [-W#warnings]
#90 0.300     1 | #warning redirecting incorrect #include <sys/poll.h> to <poll.h>
#90 0.300       |  ^
#90 0.307 /root/php_wasm.c:117:23: error: too few arguments provided to function-like macro invocation
#90 0.307   117 |                 RETURN_STRING(result);
#90 0.307       |                                     ^
#90 0.307 Zend/zend_API.h:639:9: note: macro 'RETURN_STRING' defined here
#90 0.307   639 | #define RETURN_STRING(s, duplicate)     { RETVAL_STRING(s, duplicate); return; }
#90 0.307       |         ^
#90 0.308 /root/php_wasm.c:117:3: error: use of undeclared identifier 'RETURN_STRING'
#90 0.308   117 |                 RETURN_STRING(result);
#90 0.308       |                 ^
#90 0.312 1 warning and 2 errors generated.
#90 0.314 emcc2: error: '/root/emsdk/upstream/bin/clang -target wasm32-unknown-emscripten -fignore-exceptions -fvisibility=default -mllvm -combiner-global-alias-analysis=false -mllvm -enable-emscripten-sjlj -mllvm -disable-lsr -DEMSCRIPTEN -Werror=implicit-function-declaration --sysroot=/root/emsdk/upstream/emscripten/cache/sysroot -Xclang -iwithsysroot/include/fakesdl -Xclang -iwithsysroot/include/compat -O3 -I. -Iext -Iext/json -IZend -Imain -ITSRM/ -I/root/lib/include -I/root/zlib -I/root/libzip -DWITH_CLI_SAPI=1 -I/root/libxml2 -Iext/pdo_sqlite -I/root/zlib -I/root/libpng16 -g0 /root/php_wasm.c -c -o /tmp/emscripten_temp_b8ib3nwr/php_wasm_0.o' failed (returned 1)
#90 ERROR: process "/bin/bash -c source /root/emsdk/emsdk_env.sh &&     export EXPORTED_FUNCTIONS=$'[\"_php_wasm_init\", \\n\"_phpwasm_destroy_uploaded_files_hash\", \\n\"_phpwasm_init_uploaded_files_hash\", \\n\"_phpwasm_register_uploaded_file\", \\n\"_wasm_set_phpini_path\", \\n\"_wasm_set_phpini_entries\", \\n\"_wasm_add_SERVER_entry\", \\n\"_wasm_add_uploaded_file\", \\n\"_wasm_sapi_handle_request\", \\n\"_wasm_set_content_length\", \\n\"_wasm_set_content_type\", \\n\"_wasm_set_cookies\", \\n\"_wasm_set_path_translated\", \\n\"_wasm_set_php_code\", \\n\"_wasm_set_query_string\", \\n\"_wasm_set_request_body\", \\n\"_wasm_set_request_host\", \\n\"_wasm_set_request_method\", \\n\"_wasm_set_request_port\", \\n\"_wasm_set_request_uri\", \\n\"_wasm_set_skip_shebang\" '\"$(cat /root/.EXPORTED_FUNCTIONS)\"']';     emcc -O3     --js-library /root/phpwasm-emscripten-library.js     -I .      -I ext       -I ext/json       -I Zend      -I main      -I TSRM/     -I /root/lib/include     -L/root/lib -L/root/lib/lib/     $(cat /root/.emcc-php-wasm-flags)     -o /root/output/php.js     -s EXPORTED_FUNCTIONS=\"$EXPORTED_FUNCTIONS\"     -s EXTRA_EXPORTED_RUNTIME_METHODS='[\"ccall\", \"UTF8ToString\", \"lengthBytesUTF8\", \"FS\", \"PROXYFS\"]'     -s INITIAL_MEMORY=1024MB     -s ALLOW_MEMORY_GROWTH=1             -s ASSERTIONS=0                      -s ERROR_ON_UNDEFINED_SYMBOLS=0      -s NODEJS_CATCH_EXIT=0               -s INVOKE_RUN=0                      -s EXIT_RUNTIME=1                        /root/lib/libphp.a         /root/php_wasm.c         $(cat /root/.emcc-php-wasm-sources)     -s ENVIRONMENT=$EMSCRIPTEN_ENVIRONMENT     -s FORCE_FILESYSTEM=1     -s EXPORT_NAME=\"'PHPLoader'\"" did not complete successfully: exit code: 1
------
 > [emscripten-php 46/50] RUN source /root/emsdk/emsdk_env.sh &&     export EXPORTED_FUNCTIONS=$'["_php_wasm_init", \n"_phpwasm_destroy_uploaded_files_hash", \n"_phpwasm_init_uploaded_files_hash", \n"_phpwasm_register_uploaded_file", \n"_wasm_set_phpini_path", \n"_wasm_set_phpini_entries", \n"_wasm_add_SERVER_entry", \n"_wasm_add_uploaded_file", \n"_wasm_sapi_handle_request", \n"_wasm_set_content_length", \n"_wasm_set_content_type", \n"_wasm_set_cookies", \n"_wasm_set_path_translated", \n"_wasm_set_php_code", \n"_wasm_set_query_string", \n"_wasm_set_request_body", \n"_wasm_set_request_host", \n"_wasm_set_request_method", \n"_wasm_set_request_port", \n"_wasm_set_request_uri", \n"_wasm_set_skip_shebang" '"$(cat /root/.EXPORTED_FUNCTIONS)"']';     emcc -O3     --js-library /root/phpwasm-emscripten-library.js     -I .      -I ext       -I ext/json       -I Zend      -I main      -I TSRM/     -I /root/lib/include     -L/root/lib -L/root/lib/lib/     $(cat /root/.emcc-php-wasm-flags)     -o /root/output/php.js     -s EXPORTED_FUNCTIONS="$EXPORTED_FUNCTIONS"     -s EXTRA_EXPORTED_RUNTIME_METHODS='["ccall", "UTF8ToString", "lengthBytesUTF8", "FS", "PROXYFS"]'     -s INITIAL_MEMORY=1024MB     -s ALLOW_MEMORY_GROWTH=1             -s ASSERTIONS=0                      -s ERROR_ON_UNDEFINED_SYMBOLS=0      -s NODEJS_CATCH_EXIT=0               -s INVOKE_RUN=0                      -s EXIT_RUNTIME=1                        /root/lib/libphp.a         /root/php_wasm.c         $(cat /root/.emcc-php-wasm-sources)     -s ENVIRONMENT=node     -s FORCE_FILESYSTEM=1     -s EXPORT_NAME="'PHPLoader'":
0.307   117 |                 RETURN_STRING(result);
0.307       |                                     ^
0.307 Zend/zend_API.h:639:9: note: macro 'RETURN_STRING' defined here
0.307   639 | #define RETURN_STRING(s, duplicate)     { RETVAL_STRING(s, duplicate); return; }
0.307       |         ^
0.308 /root/php_wasm.c:117:3: error: use of undeclared identifier 'RETURN_STRING'
0.308   117 |                 RETURN_STRING(result);
0.308       |                 ^
0.312 1 warning and 2 errors generated.
0.314 emcc2: error: '/root/emsdk/upstream/bin/clang -target wasm32-unknown-emscripten -fignore-exceptions -fvisibility=default -mllvm -combiner-global-alias-analysis=false -mllvm -enable-emscripten-sjlj -mllvm -disable-lsr -DEMSCRIPTEN -Werror=implicit-function-declaration --sysroot=/root/emsdk/upstream/emscripten/cache/sysroot -Xclang -iwithsysroot/include/fakesdl -Xclang -iwithsysroot/include/compat -O3 -I. -Iext -Iext/json -IZend -Imain -ITSRM/ -I/root/lib/include -I/root/zlib -I/root/libzip -DWITH_CLI_SAPI=1 -I/root/libxml2 -Iext/pdo_sqlite -I/root/zlib -I/root/libpng16 -g0 /root/php_wasm.c -c -o /tmp/emscripten_temp_b8ib3nwr/php_wasm_0.o' failed (returned 1)
------
Dockerfile:912
--------------------
 911 |     COPY ./build-assets/phpwasm-emscripten-library.js /root/phpwasm-emscripten-library.js
 912 | >>> RUN source /root/emsdk/emsdk_env.sh && \
 913 | >>>     export EXPORTED_FUNCTIONS=$'["_php_wasm_init", \n\
 914 | >>> "_phpwasm_destroy_uploaded_files_hash", \n\
 915 | >>> "_phpwasm_init_uploaded_files_hash", \n\
 916 | >>> "_phpwasm_register_uploaded_file", \n\
 917 | >>> "_wasm_set_phpini_path", \n\
 918 | >>> "_wasm_set_phpini_entries", \n\
 919 | >>> "_wasm_add_SERVER_entry", \n\
 920 | >>> "_wasm_add_uploaded_file", \n\
 921 | >>> "_wasm_sapi_handle_request", \n\
 922 | >>> "_wasm_set_content_length", \n\
 923 | >>> "_wasm_set_content_type", \n\
 924 | >>> "_wasm_set_cookies", \n\
 925 | >>> "_wasm_set_path_translated", \n\
 926 | >>> "_wasm_set_php_code", \n\
 927 | >>> "_wasm_set_query_string", \n\
 928 | >>> "_wasm_set_request_body", \n\
 929 | >>> "_wasm_set_request_host", \n\
 930 | >>> "_wasm_set_request_method", \n\
 931 | >>> "_wasm_set_request_port", \n\
 932 | >>> "_wasm_set_request_uri", \n\
 933 | >>> "_wasm_set_skip_shebang" '"$(cat /root/.EXPORTED_FUNCTIONS)"']'; \
 934 | >>>     emcc -O3 \
 935 | >>>     --js-library /root/phpwasm-emscripten-library.js \
 936 | >>>     -I .  \
 937 | >>>     -I ext   \
 938 | >>>     -I ext/json   \
 939 | >>>     -I Zend  \
 940 | >>>     -I main  \
 941 | >>>     -I TSRM/ \
 942 | >>>     -I /root/lib/include \
 943 | >>>     -L/root/lib -L/root/lib/lib/ \
 944 | >>>     $(cat /root/.emcc-php-wasm-flags) \
 945 | >>>     -o /root/output/php.js \
 946 | >>>     -s EXPORTED_FUNCTIONS="$EXPORTED_FUNCTIONS" \
 947 | >>>     -s EXTRA_EXPORTED_RUNTIME_METHODS='["ccall", "UTF8ToString", "lengthBytesUTF8", "FS", "PROXYFS"]' \
 948 | >>>     -s INITIAL_MEMORY=1024MB \
 949 | >>>     -s ALLOW_MEMORY_GROWTH=1         \
 950 | >>>     -s ASSERTIONS=0                  \
 951 | >>>     -s ERROR_ON_UNDEFINED_SYMBOLS=0  \
 952 | >>>     -s NODEJS_CATCH_EXIT=0           \
 953 | >>>     -s INVOKE_RUN=0                  \
 954 | >>>     -s EXIT_RUNTIME=1                \
 955 | >>>         /root/lib/libphp.a \
 956 | >>>         /root/php_wasm.c \
 957 | >>>         $(cat /root/.emcc-php-wasm-sources) \
 958 | >>>     -s ENVIRONMENT=$EMSCRIPTEN_ENVIRONMENT \
 959 | >>>     -s FORCE_FILESYSTEM=1 \
 960 | >>>     -s EXPORT_NAME="'PHPLoader'"
 961 |         # Emscripten complains it can't find some Asyncify functions
--------------------
ERROR: failed to solve: process "/bin/bash -c source /root/emsdk/emsdk_env.sh &&     export EXPORTED_FUNCTIONS=$'[\"_php_wasm_init\", \\n\"_phpwasm_destroy_uploaded_files_hash\", \\n\"_phpwasm_init_uploaded_files_hash\", \\n\"_phpwasm_register_uploaded_file\", \\n\"_wasm_set_phpini_path\", \\n\"_wasm_set_phpini_entries\", \\n\"_wasm_add_SERVER_entry\", \\n\"_wasm_add_uploaded_file\", \\n\"_wasm_sapi_handle_request\", \\n\"_wasm_set_content_length\", \\n\"_wasm_set_content_type\", \\n\"_wasm_set_cookies\", \\n\"_wasm_set_path_translated\", \\n\"_wasm_set_php_code\", \\n\"_wasm_set_query_string\", \\n\"_wasm_set_request_body\", \\n\"_wasm_set_request_host\", \\n\"_wasm_set_request_method\", \\n\"_wasm_set_request_port\", \\n\"_wasm_set_request_uri\", \\n\"_wasm_set_skip_shebang\" '\"$(cat /root/.EXPORTED_FUNCTIONS)\"']';     emcc -O3     --js-library /root/phpwasm-emscripten-library.js     -I .      -I ext       -I ext/json       -I Zend      -I main      -I TSRM/     -I /root/lib/include     -L/root/lib -L/root/lib/lib/     $(cat /root/.emcc-php-wasm-flags)     -o /root/output/php.js     -s EXPORTED_FUNCTIONS=\"$EXPORTED_FUNCTIONS\"     -s EXTRA_EXPORTED_RUNTIME_METHODS='[\"ccall\", \"UTF8ToString\", \"lengthBytesUTF8\", \"FS\", \"PROXYFS\"]'     -s INITIAL_MEMORY=1024MB     -s ALLOW_MEMORY_GROWTH=1             -s ASSERTIONS=0                      -s ERROR_ON_UNDEFINED_SYMBOLS=0      -s NODEJS_CATCH_EXIT=0               -s INVOKE_RUN=0                      -s EXIT_RUNTIME=1                        /root/lib/libphp.a         /root/php_wasm.c         $(cat /root/.emcc-php-wasm-sources)     -s ENVIRONMENT=$EMSCRIPTEN_ENVIRONMENT     -s FORCE_FILESYSTEM=1     -s EXPORT_NAME=\"'PHPLoader'\"" did not complete successfully: exit code: 1
file:///Users/e/wordpress-playground/packages/php-wasm/compile/build.js:211
			else reject(new Error(`Process exited with code ${code}`));
```

Reply:
I'll run the compiles for each PHP version individually tomorrow and report back for which ones it failed.

Reply:
it looks like these macros [were changed](https://github.com/php/php-src/commit/f4cfaf36e23ca47da3e352e1c60909104c059647#diff-bb1285146d0d0b77580bccc637afc7da6446eb0d3ad06925cc3a056c53de8a98L634-L636) at some point, and previous versions don't have a single-argument call.

previously I think we're looking at `RETURN_STRING(s, duplicate)` or `RETURN_STRING(s, length, duplicate)` where I think that `duplicate` is a `1` or a `0` indicating if the string value should be copied when returned or if the original string value itself should be returned.

if this is the case I bet you could add for earlier versions of PHP the following:
```
if (result ! NULL) {
#if PHP_MAJOR_VERSION >= 8
	RETURN_STRING(result);
#else
	RETURN_STRING(result, 0);
#endif
} else {
	RETURN_NULL();
}
```

Reply:
I'll compile each version separately and see where it errors

Reply:
Do you know why `packages/php-wasm/node/public/php_8_2.js` and .wasm are such big diffs?

Reply:
I'm assuming it's because of different `npm` packages on your system vs. what was last checked in. it looks like that's from emscripten doing something differently, mostly in a different order than before.

Reply:
Isn't that what `package-lock.json` is supposed to solve? 🤔

Reply:
Maybe it's also Docker env. Just guessing. It's a bit problematic for people trying to contribute. Could it be moved to build job on GH actions instead of having to run it locally?

Reply:
Everything down to 7.0 compliled fine. Trying 5.6 now.

Reply:
🙈 Been running it with the condition on

Reply:
Ok, so we only have to change it if lower than PHP 7. Updated.

Reply:
> Isn't that what package-lock.json is supposed to solve?

absolutely, but I think it doesn't always work, especially if someone somehow ends up running `npm install` instead of `npm ci`

> Everything down to 7.0 compliled fine. Trying 5.6 now.

yeah I should have been stronger in my comment above. based on what I read in the source code, I was fairly sure all we needed to do was switch to the old macro when compiling older versions of PHP